autopush/db.py

```diff
@@ -175,6 +207,113 @@ def delete_notification(self, uaid, chid, version=None):
+    def register_channel(self, uaid, channel_id):
+        """Register a channel for a given uaid"""
+        conn = self.table.connection
+        db_key = self.encode({"uaid": uaid, "chidmessageid": " "})
+        # Generate our update expression
+        expr = "ADD chids :channel_id"
+        expr_values = self.encode({":channel_id": set([channel_id])})
+        conn.update_item(
+            self.table.table_name,
+            db_key,
+            update_expression=expr,
+            expression_attribute_values=expr_values,
+        )
+        return True
+
+    @track_provisioned
+    def unregister_channel(self, uaid, channel_id):
+        """Remove a channel registration for a given uaid"""
+        conn = self.table.connection
+        db_key = self.encode({"uaid": uaid, "chidmessageid": " "})
+        expr = "DELETE chids :channel_id"
+        expr_values = self.encode({":channel_id": set([channel_id])})
+        conn.update_item(
+            self.table.table_name,
+            db_key,
+            update_expression=expr,
+            expression_attribute_values=expr_values,
+        )
+        return True
+
+    @track_provisioned
+    def all_channels(self, uaid):
+        """Retrieve a list of all channels for a given uaid"""
+        try:
+            result = self.table.get_item(consistent=True, uaid=uaid,
+                                         chidmessageid=" ")
+            return result["chids"]
+        except ItemNotFound:
+            return set([])
+
+    @track_provisioned
+    def store_message(self, uaid, channel_id, data, headers, message_id):
+        """Stores a message in the message table for the given uaid/channel with
+        the message id"""
+        self.table.put_item(data=dict(
+            uaid=uaid,
+            chidmessageid="%s:%s" % (channel_id, message_id),
+            data=data,
+            headers=headers,
+        ))
+        return True
+
+    @track_provisioned
+    def delete_message(self, uaid, channel_id, message_id):
+        """Deletes a specific message"""
+        self.table.delete_item(uaid=uaid, chidmessageid="%s:%s" % (
+            channel_id, message_id))
+        return True
+
+    def delete_messages(self, uaid, chidmessageids):
+        with self.table.batch_write() as batch:
+            for chidmessageid in chidmessageids:
+                batch.delete_item(
+                    uaid=uaid,
+                    chidmessageid=chidmessageid
+                )
+
+    @track_provisioned
+    def delete_messages_for_channel(self, uaid, channel_id):
+        """Deletes all messages for a uaid/channel_id"""
+        results = self.table.query_2(
+            uaid__eq=uaid,
+            chidmessageid__beginswith="%s:" % channel_id,
+            consistent=True,
+            attributes=("chidmessageid",),
+        )
+        chidmessageids = []
+        for item in results:
+            chidmessageids.append(item["chidmessageid"])
+            if len(chidmessageids) == 25:
```

Comment:
Oh, I had no idea DynamoDB limits batches to 25 operations. Could you add a quick comment here, please?